### PR TITLE
Critical fixes and 'legacy i2c' support

### PR DESCRIPTION
**Pull request by eflukx**
Hi, please have a look at these changes. I needed to do the alterations to get this lib going. Expect more to come :)

### CODE REVIEW

**Comment by eflukx** [comment]
Oh what I forgot to mention in the commit message (probably there's a bit too much going on in this one-commit), but there was also a bug in the address calculation in the `fn address_with_register(&self, register: u16) -> u8` (at least on _'my machine'_ using nrf-hal).

In my case the i2c address given to the i2c hal instance is the 7-bit address, without the R/W bit, where your implementation assumed an 8 bit address (so _including_ the R/W bit). This could very well be something that differs between i2c hal implementations and something to be aware of... If both approaches exist in the field (which would suck btw.) this needs to be handled somehow (e.g. in the type of the `I2cInterface` I would guess)

**Comment by eflukx** [comment]
I've been reading up on my i2c addressing implementations and this is actually handles in the embedded hal by being generic over `AddressMode`. Currently `SevenBitAddress` and `TenBitAddress` are implemented, so imo HALs implementing 8-bit i2c adressing can be considered buggy. Still if these exist _in the field_ these probably need to be supported anyway..

**Comment by markus-k** [comment]
Thank you very much for the PR! I haven't received my hardware yet, so all the code is untested so far. Especially I'm not happy with the configuration stuff yet, so that'll probably change quite a bit.

I'll have a look at your changes asap and give feedback or merge them :)

File: src/lib.rs [hunk 1]
```diff
@@ -490,29 +501,31 @@ pub trait PwmAccess<T> {
...
 
-    fn set_pwm(&mut self, start: u16, values: &[u16]) -> Result<(), Self::Error> {
-        let mut registers = [0; 396];
+    fn set_pwm(&mut self, start_dot: u16, values: &[u16]) -> Result<(), Self::Error> {
+        let mut buffer = [0; Variant0::NUM_DOTS as usize * 2];
 
-        if values.len() + start as usize > (DV::NUM_DOTS as usize) {
+        if values.len() + start_dot as usize > (DV::NUM_DOTS as usize) {
+            // TODO: probably we don't want to panic in an embedded system...
             panic!("Too many values supplied for given start and device variant.");
```
Comment (markus-k) [inline src/lib.rs:516]: I think we could just ignore any excess data here (or return an Error), but in either way this behaviour should be documented.

File: src/lib.rs [hunk 2]
```diff
@@ -490,29 +501,31 @@ pub trait PwmAccess<T> {
 impl<DV: DeviceVariant, I: RegisterAccess> PwmAccess<u16> for Lp586x<DV, I, DataMode16Bit> {
     type Error = I::Error;
 
-    fn set_pwm(&mut self, start: u16, values: &[u16]) -> Result<(), Self::Error> {
-        let mut registers = [0; 396];
+    fn set_pwm(&mut self, start_dot: u16, values: &[u16]) -> Result<(), Self::Error> {
+        let mut buffer = [0; Variant0::NUM_DOTS as usize * 2];
```
Comment (markus-k) [inline src/lib.rs:512]: We should document why we are wasting so much stack space here:
```suggestion
        // rust does not support making this dependent on a generic parameter (DV::NUM_DOTS),
        // so have have to just use the biggest variant here. Lots of stack space wasted in
        // most cases :(
        let mut buffer = [0; Variant0::NUM_DOTS as usize * 2];
```

I don't really have a better idea at the moment at least...

Comment (eflukx) [reply src/lib.rs:512]: Yes.. agree this is a bit wasteful, i also found you cannot use the generic here, so this seems the safest/easiest way to go for now..  (isn't this functionality hidden behind a 'const_generics' feature flag of some sort...?)

File: src/lib.rs (hunk 1 above)
Comment (eflukx) [reply src/lib.rs:516]: Yep.. but the error type is generic (and unknown in this place i guess). Anothe way to go is just to silently ignore the excess data (truncate)

Comment (markus-k) [reply src/lib.rs:516]: Ah yes, I also noticed the return type is `Self::Error`, not `Error<>` after commenting. That does not seem right, I'll fix that in the future.

File: src/lib.rs (hunk 2 above)
Comment (markus-k) [reply src/lib.rs:512]: I don't think this particular issues is hidden behind any feature flag, at least to my limited knowledge. I also couldn't find a workaround, but haven't searched for too long. It'd be best if we wouldn't need a buffer at all, but at least `embedded_hal` v1 removes the the `_iter` methods and I'm not sure if they were even implemented for most platforms.

One alternative might be using a const generic in the function signature, i.e. `fn set_pwm<const N: usize>(&mut self, start_dot: u16, values: &[u16; N])`, but that probably has it's implications as well (e.g. you can't just use `let foo = [0; {N*2}]` without the nightly feature `generic_const_exprs`).

Comment (eflukx) [reply src/lib.rs:512]: I think the waste is limited. When on a (very) resource constrained platform, you can always revert to the 8-bit mode (also implemented in this PR) that doesn't not create an extra intermediate buffer. On the other hand, whilst the `*Iter` methods are nice, for instance on my platform (nrf) i2c is done using DMA, for which a buffer in RAM is needed in any case (I don't see how this iter approach would work in the DMA world..)
Lochily more modern platforms typically have some kb's of memory to spare..

File: src/configuration.rs [hunk 3]
```diff
@@ -56,4 +56,64 @@ impl Configuration {
...
+        } | self.data_ref_mode.register_value()
+            << BitFlags::DEV_INITIAL_DATA_REF_MODE_SHIFT
+            | (self.max_line_num & BitFlags::DEV_INITIAL_MAX_LINE_NUM_MASK)
+                << BitFlags::DEV_INITIAL_MAX_LINE_NUM_SHIFT;
+
+        dev_initial
+    }
+
+    pub fn dev_config1_reg_value(&self) -> u8 {
+        let dev_config1 = if self.cs_turn_on_delay {
```
Comment (markus-k) [inline src/configuration.rs:73]: Looks like my changes to use `then_some(..).unwarp_or(0)` have gone missing here.

Comment (eflukx) [reply src/configuration.rs:73]: Right.. will have a look.. I was working on this at the same time you pushed the support for cfg reg 2 and 3. So took your work and spliced it in there.. possibly missed something there.. :)

**Comment by eflukx** [comment]
As a sidenote: I have my device working with embedded graphics, of which I [shared the initial code here](https://github.com/eflukx/lp586x-rs/tree/embedded-graphics)

My display consists of 2 lp5864's for upper/lower part of the matrix display (made so due to parts shortage at the time of ordering). So there's a hacky approach to do a compound display (for my use case fixed to a 1x2 display). 
Creating a real flexible compound display system is a pain to create without `box dyn` (and `alloc`), so went with a "fits-my-use-case-only" approach. Don't think this has any place in the main lib, but embedded graphics support _has_ imo. (and this may be a nice starter)

**Comment by markus-k** [comment]
Interersting! I think you copied the wrong link though ;) But I found the branch in your fork.

I have a PCB in the mail with 5 LP5864s on there (for 72 RGBWW LEDs = 360 dots), using SPI.  So it's good to have someone actually use the I2C part as well. The LEDs aren't arranged in a grid-pattern, so `embedded-graphics` won't be of much use for be personally, but I thought about it as well and it's certainly a good idea to support it. Looking forward to it!

I think I'll merge this for now, happy to accept more PRs! One thing I'll certainly want to sort out is improved unit tests (first of all, tests that don't fail). Once I feel like the library is actually usable and tested on real hardware I'll publish it on crates.io.

**Pull request merged**